src/services/download/gcs.rs

```diff
@@ -109,117 +105,106 @@ fn request_new_token(
-                        )
-                            as Box<dyn Stream<Item = _, Error = _>>))
-                    } else {
-                        log::trace!(
-                            "Unexpected status code from GCS {} (from {}): {}",
-                            &key,
-                            source.bucket,
-                            response.status()
-                        );
-                        Ok(None)
-                    }
-                }
-                Err(e) => {
-                    log::trace!(
-                        "Skipping response from GCS {} (from {}): {} ({:?})",
-                        &key,
-                        source.bucket,
-                        &e,
-                        &e
-                    );
-                    Ok(None)
-                }
-            })
-    };
-
-    let response = Retry::spawn(
-        ExponentialBackoff::from_millis(10).map(jitter).take(3),
-        try_response,
+    let url = format!(
+        "https://www.googleapis.com/download/storage/v1/b/{}/o/{}?alt=media",
+        percent_encode(source.bucket.as_bytes(), PATH_SEGMENT_ENCODE_SET),
+        percent_encode(key.as_bytes(), PATH_SEGMENT_ENCODE_SET),
     );
 
-    let stream = Box::new(response.map_err(|e| match e {
-        tokio_retry::Error::OperationError(e) => e,
-        e => panic!("{}", e),
-    }));
-    super::download_future_stream(
-        SourceFileId::Gcs(source, download_path),
-        stream,
-        destination,
-    )
+    let mut backoff = ExponentialBackoff::from_millis(10).map(jitter).take(3);
+    let response = loop {
+        let result = start_request(&url, &token).await;
+        match backoff.next() {
+            Some(duration) if result.is_err() => delay(duration).await,
+            _ => break result,
+        }
+    };
+
+    match response {
+        Ok(response) => {
+            if response.status().is_success() {
+                log::trace!("Success hitting GCS {} (from {})", &key, source.bucket);
+                let stream = response
+                    .payload()
+                    .compat()
+                    .map(|i| i.map_err(|e| e.context(DownloadErrorKind::Io).into()));
+                super::download_stream(
+                    SourceFileId::Gcs(source, download_path),
+                    stream,
+                    destination,
+                )
+                .await
+            } else {
+                log::trace!(
+                    "Unexpected status code from GCS {} (from {}): {}",
+                    &key,
+                    source.bucket,
+                    response.status()
+                );
+                Ok(DownloadStatus::NotFound)
+            }
+        }
+        Err(e) => {
+            log::trace!(
+                "Skipping response from GCS {} (from {}): {} ({:?})",
+                &key,
+                source.bucket,
+                &e,
+                &e
```

Comment:
Is there a reason we format the same error twice? Looks like once should be enough. Also, no need to `&` all these, as the macro will take references only.